plugins/rebalance/rest.go

```diff
@@ -0,0 +1,323 @@
+		return
+	}
+
+	if vol.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotStarted)
+		return
+	}
+
+	if vol.DistCount == 1 {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotDistribute)
+		return
+	}
+
+	// Check for remove- brick pending
+	//TODO
+
+	// A simple transaction to start rebalance on all nodes
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err)
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebalance-start",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebalance-store",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		unlock,
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	rebalinfo := createRebalanceInfo(volname, &req)
+	if rebalinfo == nil {
+		logger.WithError(err).Error("failed to create Rebalance info")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Ctx.Set("rinfo", rebalinfo)
+	if err != nil {
+		logger.WithError(err).Error("failed to set rebalance info in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		/* TODO: Need to handle failure case. Unlike other process,
+		 * rebalance process is one per node and depends on volfile change.
+		 * Need to handle scenarios where process is started in
+		 * few nodes and failed in few others */
+		logger.WithFields(log.Fields{
+			"error":   err.Error(),
+			"volname": volname,
+		}).Error("failed to start rebalance on volume")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	rebalinfo, err = GetRebalanceInfo(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		return
+	}
+
+	logger.WithField("volname", rebalinfo.Volname).Info("rebalance started")
+
+//TODO: Fix this!
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, rebalinfo)
+}
+
+
+func rebalanceStopHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+
+	// Validate rebalance command
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		return
+	}
+
+	rebalinfo, err := GetRebalanceInfo(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		return
+	}
+
+	// Check rebalance process is started or not
+	if rebalinfo.Status != rebalanceapi.Started {
```

Comment:
you can also add `rebalinfo.Status != rebalanceapi.Completed`

Reply:
That could also mean that the process failed , in which case it will also not running. So we usually only allow a stop if the status == started.